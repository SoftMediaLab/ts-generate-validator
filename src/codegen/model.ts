import { Message } from '../localization/model';
import { UserContext, ValidationConfig } from '../config/model';

export type Data = Record<string, any>;

export type GeneratedValidation = <D extends Data, C extends UserContext>(
  data: D,
  config?: ValidationConfig,
  context?: C
) => void | Promise<void>;

export interface ValidationError<D extends Data> {
  field: keyof D | (keyof D)[];
  message: Message;
}

export interface ValidationException<D extends Data> {
  errors: ValidationError<D>[];
}
