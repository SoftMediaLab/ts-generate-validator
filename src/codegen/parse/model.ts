import { ValidationType } from '../../validators/model';
import { Decorator } from 'ts-file-parser';

export interface ClassFieldTypeMetadata {
  validationType: ValidationType;
  referencePath?: string;
  name?: string;
}

export interface ClassFieldMetadata {
  name: string;
  type: ClassFieldTypeMetadata;
  optional: boolean;
  decorators: Decorator[];
}

export interface ClassMetadata {
  name: string;
  decorators: Decorator[];
  fields: ClassFieldMetadata[];
}

export interface InputFileMetadata {
  name: string;
  classes: ClassMetadata[];
}

export interface CustomTypeEntry {
  fileIndex: number;
  classIndex: number;
  fieldIndex: number;
}

export type EnumDictionary = { [file: string]: string[] };
