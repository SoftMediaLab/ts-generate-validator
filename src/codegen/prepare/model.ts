export interface PreparedImport {
  clauses: string;
  path: string;
}

export type PreparedImportMap = {
  [path: string]: {
    [clause: string]: boolean;
  };
};

export interface PreparedValidatorPayloadItem {
  property: string;
  value: string;
  type: string;
}

export interface PreparedValidationItem {
  propertyName: string;
  validatorName: string;
  validatorPayload: PreparedValidatorPayloadItem[];
}

export interface PreparedValidation {
  async: boolean;
  name: string;
  modelClassName: string;
  items: PreparedValidationItem[];
}

export type PreparedDataItem = {
  filePath: string;
  fileName: string;
  imports: PreparedImport[];
  validationArgs: Record<string, string>;
  validations: PreparedValidation[];
};
