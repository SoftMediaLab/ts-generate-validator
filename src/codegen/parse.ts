import { parseStruct, Decorator, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
// import * as path from 'path';
// import { ArrayType, BasicType } from 'ts-file-parser';

export interface DecoratedClassFieldMetadata {
  name: string;
  type: any;
  optional: any;
  decorators: Decorator[];
}

export interface DecoratedClassMetadata {
  name: string;
  decorators: Decorator[];
  fields: DecoratedClassFieldMetadata[];
}

export interface InputFileMetadata {
  name: string;
  imports: string[];
  classes: DecoratedClassMetadata[];
}

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  return files.map((file) => {
    console.log(`Reading ${file}...`);
    const content = fs.readFileSync(file, 'utf-8');
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const structure = parseStruct(content, {}, file);

    //if (file.match(/^.*common.ts$/)) {
    //console.log(JSON.stringify(structure, null, 4));
    //}

    return {
      name: structure.name.replace(/\/+/g, '.'),
      imports: buildImports(structure._imports),
      classes: []
    };
  });
};

const buildImports = (importNodes: ImportNode[]): string[] => {
  return importNodes.map((iNode) => {
    return `import { ${iNode.clauses.join(', ')} } from ${iNode.absPathNode.join('/')}`;
  });
};
