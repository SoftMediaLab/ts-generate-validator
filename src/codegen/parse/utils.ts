import { primitiveValidationTypes } from './../../validators/model';
import { ValidationType } from '../../validators/model';

export const getValidationTypeByTypeName = (typeName: string): ValidationType => {
  if (primitiveValidationTypes.includes(typeName as any)) {
    return <ValidationType>typeName;
  }

  const notSupportedTypes = ['null', 'undefined', 'symbol', 'function', 'object'];
  if (notSupportedTypes.includes(typeName)) {
    return ValidationType.notSupported;
  }

  return ValidationType.unknown;
};
