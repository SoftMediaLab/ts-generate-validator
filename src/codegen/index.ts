import { parseInputFiles } from './parse';
import { getAllFiles } from './utils/getAllFiles';
import { getCodegenConfig } from './../config/codegen';

export const createValidators = (): void => {
  const config = getCodegenConfig();
  const inputFiles = getAllFiles(config.inputPath);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const inputFilesMetadata = parseInputFiles(inputFiles);

  // TODO: implement
};
