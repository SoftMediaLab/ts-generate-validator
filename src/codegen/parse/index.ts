import { cutFileExt, hasFileExt, isPackagePath, normalizeFileExt, normalizePath } from './../../utils/path';
import { IgnoreValidation } from './../../decorators/index';
import { IssueError, ErrorInFile, outWarning } from './../utils/error';
import { ValidationType } from './../../validators/model';
import { buildClassesMetadata } from './classes';
import { parseStruct, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
import * as path from 'path';
import {
  InputFileMetadata,
  ClassFieldTypeMetadata,
  ClassFieldBasicTypeMetadata,
  CustomTypeEntry,
  EnumDictionary,
  ImportMetadata,
  InputFileDesc,
  ClassMetadata
} from './model';

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  const inputFiles: InputFileDesc[] = files.map((f) => ({ path: f, parseClasses: true }));
  const metadata: InputFileMetadata[] = [];
  const customTypeEntries: CustomTypeEntry[] = [];
  const enumDictionary: EnumDictionary = {};

  do {
    const inputFileDesc = inputFiles.shift();

    if (!inputFileDesc) {
      break;
    }

    const inputFilePath = path.resolve(process.cwd(), normalizeFileExt(inputFileDesc.path));
    let content: string;
    try {
      content = fs.readFileSync(inputFilePath, 'utf-8');
    } catch (err) {
      outWarning(`Failed to read input file "${inputFilePath}" -> skip (${err.message})`);
      continue;
    }
    const structure = parseStruct(content, {}, inputFileDesc.path);

    if (structure.enumDeclarations.length) {
      enumDictionary[inputFileDesc.path] = structure.enumDeclarations.map((decl) => decl.name);
    }

    const imports = buildImportsMetadata(structure._imports);
    imports.forEach((imp) => {
      const isAlreadyParsed = metadata.some((m) => m.name === imp.absPath);
      if (inputFileDesc.parseClasses && !isAlreadyParsed && !isPackagePath(imp.absPath)) {
        inputFiles.push({ path: imp.absPath, parseClasses: false });
      }
    });

    const classes: ClassMetadata[] = [];
    if (inputFileDesc.parseClasses && structure.classes.length) {
      const classesMetadata = buildClassesMetadata(metadata.length, structure.classes, structure._imports);
      if (classesMetadata.classesForValidation.length) {
        customTypeEntries.push(...classesMetadata.customTypeEntries);
        classes.push(...classesMetadata.classesForValidation);
      }
    }

    metadata.push({
      name: structure.name,
      classes,
      imports,
      functions: structure.functions.map(({ name, isExport, isAsync }) => ({ name, isExported: isExport, isAsync }))
    });
  } while (inputFiles.length);

  resolveCustomTypes({ metadata, enumDictionary, customTypeEntries });
  validateNestedClasses({ metadata, customTypeEntries });

  return metadata;
};

export const resolveCustomTypes = ({
  metadata,
  enumDictionary,
  customTypeEntries
}: {
  metadata: InputFileMetadata[];
  enumDictionary: EnumDictionary;
  customTypeEntries: CustomTypeEntry[];
}): void => {
  customTypeEntries.forEach(({ fileIndex, classIndex, fieldIndex }) => {
    const fieldTypeMetadata = metadata[fileIndex].classes[classIndex].fields[fieldIndex].type;

    const resolveBasicType = (typeMetadata: ClassFieldBasicTypeMetadata): void => {
      if (typeMetadata.validationType !== ValidationType.unknown) {
        return;
      }

      // Trying to fix empty referencePath -> set to current file with model class
      if (!typeMetadata.referencePath) {
        typeMetadata.referencePath = metadata[fileIndex].name;
      }

      const { referencePath, name: typeName } = typeMetadata;
      if (!referencePath || !typeName) {
        typeMetadata.validationType = ValidationType.notSupported;
        return;
      }

      if (enumDictionary[referencePath]?.includes(typeName)) {
        typeMetadata.validationType = ValidationType.enum;
        return;
      }

      const metadataItemWithNestedClass = metadata.find(({ name: fileName, classes }) => {
        let fileNameMatched = true;

        const fileNameHasExt = hasFileExt(fileName);
        const refPathHasExt = hasFileExt(referencePath);

        if (fileNameHasExt === refPathHasExt) {
          fileNameMatched = fileName === referencePath;
        } else {
          fileNameMatched = cutFileExt(fileName) === cutFileExt(referencePath);
        }

        if (!fileNameMatched) {
          return false;
        }

        return Boolean(classes.find(({ name: className }) => className === typeName));
      });
      if (metadataItemWithNestedClass) {
        typeMetadata.validationType = ValidationType.nested;
        typeMetadata.referencePath = metadataItemWithNestedClass.name;
        return;
      }

      typeMetadata.validationType = ValidationType.notSupported;
    };

    const resolveAnyType = (typeMetadata: ClassFieldTypeMetadata): void => {
      if (typeMetadata.validationType === ValidationType.array) {
        return resolveAnyType(typeMetadata.arrayOf);
      }
      if (typeMetadata.validationType === ValidationType.union) {
        return typeMetadata.unionTypes.forEach(resolveAnyType);
      }
      return resolveBasicType(typeMetadata);
    };

    resolveAnyType(fieldTypeMetadata);
  });
};

export const validateNestedClasses = ({
  metadata,
  customTypeEntries
}: {
  metadata: InputFileMetadata[];
  customTypeEntries: CustomTypeEntry[];
}): void => {
  customTypeEntries.forEach(({ fileIndex, classIndex, fieldIndex }) => {
    const baseClassName = metadata[fileIndex].classes[classIndex].name;
    const baseClassPath = metadata[fileIndex].name;
    const fieldTypeMetadata = metadata[fileIndex].classes[classIndex].fields[fieldIndex].type;

    const checkBasicType = (typeMetadata: ClassFieldBasicTypeMetadata): void => {
      if (typeMetadata.validationType !== ValidationType.nested) {
        return;
      }

      const checkNestedClasses = (className: string, classPath: string) => {
        if (className === baseClassName && classPath === baseClassPath) {
          const fieldName = metadata[fileIndex].classes[classIndex].fields[fieldIndex].name;
          throw new ErrorInFile(
            `Class "${baseClassName}" has circular dependency in field "${fieldName}" with type "${typeMetadata.name}". Ð¡ircular dependencies are not allowed because they lead to infinite validation. Change field type or add "@${IgnoreValidation.name}" decorator.`,
            baseClassPath
          );
        }

        const _fileMetadata = metadata.find(({ name: filePath }) => filePath === classPath);
        if (!_fileMetadata) {
          throw new IssueError(
            `Referenced file metadata not found for nested class "${className}" (referenced file path: "${classPath}").`
          );
        }

        const _classMetadata = _fileMetadata.classes.find(({ name }) => name === className);
        if (!_classMetadata) {
          throw new IssueError(`Metadata not found for nested class "${className}" in file metadata "${classPath}").`);
        }

        _classMetadata.fields.forEach(({ type: _fieldTypeMetadata }) => {
          const _checkBasicType = (_typeMetadata: ClassFieldBasicTypeMetadata): void => {
            const { name: _name, validationType: _validationType, referencePath: _referencePath } = _typeMetadata;

            if (_validationType === ValidationType.nested && _name && _referencePath) {
              checkNestedClasses(_name, _referencePath);
            }
          };

          const _checkAnyType = (_typeMetadata: ClassFieldTypeMetadata): void => {
            if (_typeMetadata.validationType === ValidationType.array) {
              return _checkAnyType(_typeMetadata.arrayOf);
            }
            if (_typeMetadata.validationType === ValidationType.union) {
              return _typeMetadata.unionTypes.forEach(_checkAnyType);
            }
            return _checkBasicType(_typeMetadata);
          };

          _checkAnyType(_fieldTypeMetadata);
        });
      };

      if (typeMetadata.name && typeMetadata.referencePath) {
        checkNestedClasses(typeMetadata.name, typeMetadata.referencePath);
      }
    };

    const checkAnyType = (typeMetadata: ClassFieldTypeMetadata): void => {
      if (typeMetadata.validationType === ValidationType.array) {
        return checkAnyType(typeMetadata.arrayOf);
      }
      if (typeMetadata.validationType === ValidationType.union) {
        return typeMetadata.unionTypes.forEach(checkAnyType);
      }
      return checkBasicType(typeMetadata);
    };

    checkAnyType(fieldTypeMetadata);
  });
};

const buildImportsMetadata = (nodes: ImportNode[]): ImportMetadata[] => {
  return nodes.map(({ clauses, absPathString }) => {
    return {
      clauses,
      absPath: normalizePath(path.relative(process.cwd(), path.resolve(absPathString)))
    };
  });
};
