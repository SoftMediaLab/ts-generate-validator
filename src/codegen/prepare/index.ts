import { allowedFileExt } from './../model';
import { normalizePath, cutFileExt, isPackagePath } from './../../utils/path';
import { buildValidationFromClassMetadata } from './validation';
import { PreparedDataItem, PreparedValidation, PreparedImportMap, PreparedImport } from './model';
import * as path from 'path';
import { InputFileMetadata } from '../parse/model';
import { CodegenConfig } from '../../config/model';
import * as pkg from '../../../package.json';

export const prepareDataForRender = (
  inputFilesMetadata: InputFileMetadata[],
  config: CodegenConfig
): PreparedDataItem[] => {
  const preparedData: PreparedDataItem[] = [];

  inputFilesMetadata.forEach((metadata) => {
    const { name, classes, imports: inputFileImportsMetadata, functions: inputFileFunctionsMetadata } = metadata;
    const inputFilePath = cutFileExt(name);

    const filePath = buildOutputFilePath({ inputFileName: name, config });
    const fileName = buildOutputFileName(name);
    const filePathAbs = `${filePath}/${fileName}`;

    const importMap = buildBaseImportMap();
    const handleImportAdd = (targetPath: string, clause: string, isPackageName?: boolean): void => {
      const importPathAbs = path.relative(process.cwd(), targetPath);

      if (filePathAbs === importPathAbs) {
        return;
      }

      isPackageName = isPackageName ?? isPackagePath(targetPath);
      const importPath = isPackageName ? targetPath : normalizeImportPathForFile(filePath, targetPath);

      if (!importMap[importPath]) {
        importMap[importPath] = {};
      }

      importMap[importPath][clause] = true;
    };

    const validations: PreparedValidation[] = [];

    classes.forEach((cls) => {
      const validation = buildValidationFromClassMetadata({
        cls,
        clsFileName: name,
        addImport: handleImportAdd,
        inputFileImportsMetadata,
        inputFileFunctionsMetadata,
        inputFilePath,
        config
      });
      if (validation) {
        handleImportAdd(path.resolve(name), cls.name, false);
        validations.push(validation);
      }
    });

    const imports = buildImportsFromMap(importMap);

    if (validations.length) {
      preparedData.push({
        filePath,
        fileName,
        imports,
        validations
      });
    }
  });

  return preparedData;
};

const normalizeImportPathForFile = (filePath: string, importPath: string): string => {
  let resultPath = normalizePath(path.relative(filePath, importPath));
  if (!resultPath.startsWith('./')) {
    resultPath = `./${resultPath}`;
  }
  if (resultPath.match(new RegExp(`\\.(${allowedFileExt.join('|')})$`))) {
    resultPath = cutFileExt(resultPath);
  }
  return resultPath;
};

export const buildOutputFilePath = ({ config }: { inputFileName: string; config: CodegenConfig }): string => {
  let result = path.resolve(process.cwd(), config.outputPath);
  result = path.relative(process.cwd(), result);
  return normalizePath(result);
};

export const buildOutputFileName = (inputFileName: string): string => {
  const pathToInputFileFromRoot = path.relative(process.cwd(), inputFileName);
  return (
    cutFileExt(pathToInputFileFromRoot)
      .replace(/\s/g, '_')
      .replace(/[/\\]+/g, '.') + '.ts'
  );
};

const buildBaseImportMap = (): PreparedImportMap => {
  const map: PreparedImportMap = {};
  map[pkg.name] = {
    GeneratedValidation: true,
    ValidationConfig: true,
    ValidationError: true,
    ValidationException: true,
    Data: true,
    UserContext: true,
    getConfig: true,
    mergeDeep: true
  };
  return map;
};

const buildImportsFromMap = (map: PreparedImportMap): PreparedImport[] => {
  return Object.keys(map).map((path) => ({
    clauses: Object.keys(map[path])
      .filter((clause) => map[path][clause])
      .join(', '),
    path
  }));
};
