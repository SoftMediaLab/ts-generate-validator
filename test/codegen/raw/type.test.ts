import { readAllOutputFiles } from './../../utils/output';
import { removeGeneratedValidators } from '../../utils/output';
import { createValidators } from './../../../src/codegen/index';
import { prepareDataForRender } from './../../../src/codegen/prepare/index';
import { ValidationType } from '../../../src/validators/model';
import { parseInputFiles } from '../../../src/codegen/parse/index';
import * as path from 'path';
import { getAllFiles } from '../../../src/codegen/utils/getAllFiles';
import { getCodegenConfig } from '../../../src/config/codegen';
import { deleteConfigFile, createConfigFile } from '../../utils/config';

describe('codegen/raw/type', () => {
  const prevRootDir = process.cwd();

  beforeAll(() => {
    const testRootDir = process.cwd() + '/test/mock/raw/type';
    process.chdir(testRootDir);
    createConfigFile({
      inputPath: '.'
    });
  });

  afterAll(() => {
    deleteConfigFile();
    process.chdir(prevRootDir);
  });

  test('parse & prepare', () => {
    const config = getCodegenConfig();
    const inputFiles = getAllFiles(path.resolve(process.cwd(), config.inputPath));
    const inputFilesMetadata = parseInputFiles(inputFiles);

    inputFilesMetadata.forEach(({ name: fileName, classes }, index) => {
      if (index === 0) {
        expect(fileName === path.resolve('model'));

        classes.forEach(({ name: className, fields }, index) => {
          if (index === 0) {
            expect(className).toBe('ClassWithNonPrimitiveProperty');

            fields.forEach(({ name: fieldName, type, optional }) => {
              if (fieldName === 'someProperty') {
                expect(optional).toBeTruthy();
                expect(type.name).toBe('SomeNonPrimitiveStructure');
                expect(type.referencePath).toBe(fileName);
                expect(type.validationType).toBe(ValidationType.notSupported);
              }
            });
          }
          if (index === 1) {
            expect(className).toBe('ClassWithEnumProperty');

            fields.forEach(({ name: fieldName, type, optional }) => {
              if (fieldName === 'someProperty') {
                expect(optional).toBeTruthy();
                expect(type.name).toBe('SomeEnum');
                expect(type.referencePath).toBe(fileName);
                expect(type.validationType).toBe(ValidationType.enum);
              }
            });
          }
        });
      }
    });

    const dataForRender = prepareDataForRender(inputFilesMetadata, config);
    expect(dataForRender).toMatchSnapshot('data for render');
  });

  test('create validators', async () => {
    await createValidators();

    readAllOutputFiles(({ file, content }) => {
      expect(content).toMatchSnapshot(`generated validators at "${file}"`);
    });

    removeGeneratedValidators();
  });
});
