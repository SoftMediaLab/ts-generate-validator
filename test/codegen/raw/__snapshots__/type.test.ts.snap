// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen/raw/type create validators: generated validators at "generated/validators/subfolder_#1.subfolder_#2.model.ts" 1`] = `
"import {
  GeneratedValidation,
  GeneratedValidationPayload,
  UserContext,
  initConfig,
  getConfig,
  mergeDeep,
  typeValidator
} from 'ts-generate-validator';

import {
  ClassWithNonPrimitiveProperty,
  SomeEnum,
  ClassWithEnumProperty
} from './../../subfolder #1/subfolder #2/model';

import * as configFromFile from './../../ts-generate-validator-config.json';

initConfig(configFromFile);

export const classWithNonPrimitivePropertyValidator: GeneratedValidation = (
  payload: GeneratedValidationPayload<typeof ClassWithNonPrimitiveProperty, UserContext>
) => {
  // eslint-disable-next-line
  const config = mergeDeep({}, getConfig(), payload.config);
};

export const classWithEnumPropertyValidator: GeneratedValidation = (
  payload: GeneratedValidationPayload<typeof ClassWithEnumProperty, UserContext>
) => {
  const config = mergeDeep({}, getConfig(), payload.config);

  typeValidator({
    property: payload.data.someProperty,
    propertyName: 'someProperty',
    data: payload.data,
    config,

    typeDescription: SomeEnum,

    type: 'enum'
  });
};
"
`;

exports[`codegen/raw/type parse & prepare: data for render 1`] = `
Array [
  Object {
    "configFilePath": "./../../ts-generate-validator-config.json",
    "fileName": "subfolder_#1.subfolder_#2.model.ts",
    "filePath": "generated\\\\validators",
    "imports": Array [
      Object {
        "clauses": "GeneratedValidation, GeneratedValidationPayload, UserContext, initConfig, getConfig, mergeDeep, typeValidator",
        "path": "ts-generate-validator",
      },
      Object {
        "clauses": "ClassWithNonPrimitiveProperty, SomeEnum, ClassWithEnumProperty",
        "path": "./../../subfolder #1/subfolder #2/model",
      },
    ],
    "validationArgs": Object {
      "config": "config",
      "data": "data",
    },
    "validations": Array [
      Object {
        "async": false,
        "items": Array [],
        "modelClassName": "ClassWithNonPrimitiveProperty",
        "name": "classWithNonPrimitivePropertyValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "typeDescription",
                "type": "object",
                "value": "SomeEnum",
              },
              Object {
                "property": "type",
                "type": "string",
                "value": "enum",
              },
            ],
          },
        ],
        "modelClassName": "ClassWithEnumProperty",
        "name": "classWithEnumPropertyValidator",
      },
    ],
  },
]
`;
